Legacy shortcodes: block editor compatibility

barryhughes opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

In the pre-Gutenburg world, or when using the classic editor, shortcodes are typically entered in the editor but are only rendered during frontend requests. In the world of blocks, though, shortcodes may be (pre-)rendered in an admin context both when the block editor loads, and when a post is saved (which will happen via the REST API).

For this reason, any functions the shortcode handler might call need to be available, or else the shortcode handler needs to defensively check that they are available first. As described in the linked issue, there are particular problems relating to shortcodes that call our notice functions or that try to interact with the session handler (which is not instantiated for admin or REST requests).

Closes https://github.com/woocommerce/woocommerce/issues/38169.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Start by defining a shortcode. Here's some custom code you can use, which could be added to a location such as `wp-content/mu-plugins/test-40648.php` (once you finish testing, remember to remove it!):

```php
<?php
add_shortcode( 'my_test', function () {
	ob_start();
	wc_print_notices();
	echo '<div>MY CUSTOM SHORTCODE</div>';
	return ob_get_clean();
} );
```

2. Then, create a new page and add the shortcode (just to be extra clear, this should all be done using the block editor):

<div align="center"><img width="700" alt="Screenshot 2023-10-25 at 16 10 18" src="https://github.com/woocommerce/woocommerce/assets/3594411/f1863eb7-2d18-4e2c-a2e8-6d2e0f5a7cd3"></div>

3. With this branch checked out, you should be able to publish and view the page without encountering any errors. 
4. With current `trunk` checked out, you will encounter problems. For instance, if you performed the above steps using this branch first, and then switch to trunk and reload the page editor, you will encounter a fatal error:

```
Fatal error: Uncaught Error: Call to undefined function wc_print_notices()
```

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

### Notes

In an ideal world, I think we'd cover this with an E2E test. My attempts to do that so far have been very flaky, though, so I decided not to commit it. We can either move forward without, or pause and (perhaps with help from Solaris) try to get that in place, or possibly circle back and add later.

barryhughes committed:
Load standard includes and template functions in REST API and some admin contexts.

This is primarily needed to support the use of legacy shortcodes within the block editor, as pre-rendering means the shortcodes will be invoked (and therefore any frontend/template functions the shortcode handler might call need to be available).

barryhughes committed:
Check session is available before accessing its methods from within shortcodes.

barryhughes committed:
Changelog.

barryhughes committed:
Coding standards (no assignment inside a conditional).

barryhughes committed:
Whitespace.

coreymckrill commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -417,7 +418,7 @@ private function is_request( $type ) {
 				return defined( 'DOING_CRON' );
 			case 'frontend':
-				return ( ! is_admin() || defined( 'DOING_AJAX' ) ) && ! defined( 'DOING_CRON' ) && ! $this->is_rest_api_request();
+				return ( ! is_admin() || defined( 'DOING_AJAX' ) ) && ! defined( 'DOING_CRON' );
```
Kind of a nitpick, but it might be more accurate to add a separate case for `rest-api` and move `$this->is_rest_api_request()` into it. Also, defensively, it seems like `default:` should be added in a stack with `case 'frontend:` in case the front end case becomes more specific in the future.

barryhughes replied:
You're right! Updated in https://github.com/woocommerce/woocommerce/pull/40648/commits/b43a8cfc47d038b0c0c87f4873f69a70b0fc36cb.

- I restored the original test for `'frontend'` which, as you say, is more accurate.
- I initially added a case for `'rest-api'` ... then stripped it out in https://github.com/woocommerce/woocommerce/pull/40648/commits/9425dd049d81d38c76a18618144c7f1487852b77: this is a private method, so there are no external users, and internally we can simply use existing method `is_rest_api_request`.
- I did not make `'frontend'` the default clause. Though I definitely think there is a case for adding a default that returns false, it would seem risky to support a scenario where a typo like `is_request( 'badmin' )` evaluates to `is_request( 'frontend' )`.

coreymckrill replied:
> Though I definitely think there is a case for adding a default that returns false, it would seem risky to support a scenario where a typo like is_request( 'badmin' ) evaluates to is_request( 'frontend' ).

Good point 👍

coreymckrill commented on plugins/woocommerce/includes/class-woocommerce.php:
```diff
@@ -584,7 +585,10 @@ public function includes() {
+		// We load frontend includes in the post editor, because they may be invoked via pre-loading of blocks.
+		$in_post_editor = doing_action( 'load-post.php' ) || doing_action( 'load-post-new.php' );
+
+		if ( $this->is_request( 'frontend' ) || $in_post_editor ) {
```
If you add the `rest-api` case mentioned above, you could use that here instead of `$in_post_editor`. Rationale being that there could be other non-frontend cases besides the post editor where a REST API request could potentially attempt to render a shortcode.

coreymckrill replied:
On the other hand, `frontend_includes` loads 12 more files, and I'm not sure whether that would appreciably impact REST API performance or not...

coreymckrill replied:
Just going to sit here and talk to myself while doing this PR review... 😅 

I think I would err on the side of preventing fatals, and if loading those extra files during every REST API request does end up having some kind of noticeable performance impact, we could devise a workaround for that in a separate PR.

barryhughes replied:
Updated in https://github.com/woocommerce/woocommerce/pull/40648/commits/b43a8cfc47d038b0c0c87f4873f69a70b0fc36cb.

- Though I decided in https://github.com/woocommerce/woocommerce/pull/40648/commits/9425dd049d81d38c76a18618144c7f1487852b77 *not* to support a `'rest-api'` case, I have added a call to `is_rest_api_request` which amounts to the same thing.
- We cannot strip the `$in_post_editor` check. To see this element of the problem, go ahead and create a page or post with the test shortcode, publish, then remove this test and reload the editor (fatal error).

barryhughes replied:
Regarding performance, we're chiefly loading class and function definitions and are not executing bodies of logic. That should be extremely low overhead (especially with opcode caching). That said, there are indeed some side effects as a result of including those files (ie, where we call `add_action()` without anything being invoked) ... even so, there are very few of these and I still expect overhead to be very low.

coreymckrill commented on plugins/woocommerce/includes/wc-notice-functions.php:
```diff
@@ -137,7 +137,14 @@ function wc_print_notices( $return = false ) {
+
+	// If the session handler has not initialized, there will be no notices for us to read.
+	if ( null === $session ) {
+		return null;
```
I think the `@return` clause in the doc block for this is wrong. It seems like it should be `string|void`, since that's what it does when you echo the content instead of returning it. As such, this should probably just be:

```suggestion
		return;
```

barryhughes replied:
Sharp eyes, great catch! Updated in https://github.com/woocommerce/woocommerce/pull/40648/commits/4a1b225571a42b500f26f69de6d00bd412f52bba.

coreymckrill reviewed (commented):
👍 Works just as the testing instructions describe! A couple of comments/questions to ponder before merging, though.

barryhughes committed:
Support testing if the request is of type 'rest-api'.

This preserves the accuracy of the 'frontend' definition. Additionally, having tightened that up, we explicitly include REST API requests as a condition in which we should load frontend includes (needed, to support quirks in terms of how the block editor does things—such as creating FE previews of blocks).

barryhughes committed:
`rest-api` is not used, and this is a private method so it is essentially dead-code.

barryhughes committed:
Return type clean-up for `wc_print_notices()`.

coreymckrill reviewed (approved):
👍 Works great
